core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttribute.java

```diff
@@ -270,4 +272,71 @@ public void setIfAbsent(Map<ExecutionAttribute<?>, Object> attributes, T value)
             attributes.computeIfAbsent(realAttribute, k -> writeMapping.apply(null, value));
         }
     }
+
+    private static final class MappedValueStorage<T, U> implements ValueStorage<T> {
+        private final Supplier<ExecutionAttribute<U>> realAttributeSupplier;
+        private final AtomicReference<T> previousValue = new AtomicReference<>();
```

Comment:
I'm not really getting the benefit of `previousValue` here. Is it so you don't need to call `attributes.get()`?

Reply:
Chatted offline, previousValue is required because the transformation to the derived value is a lossy process, so we need to store this value separately.

Reply:
Commenting for more context:
The signer execution attributes (`AwsSignerExecutionAttributes`) now get dervied from a "real" execution attribute, `SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME`. `SELECTED_AUTH_SCHEME` holds the auth-scheme object which has info, such as signer properties, which are where the "derived" execution attributes get converted from. 

However, for `ResolvedChecksumSpecs` attribute, the conversion is from `ChecksumSpecs` to `ChecksumAlgorithm`. `ChecksumSpecs` has information like `isRequestStreaming`, `isRequestValidation`, and this info gets lost when it gets derived from `ChecksumAlgorithm`. This new `MappedAttribute` keeps the value of the attribute when it is mapped, so that the conversion back can use information that would be normally lost.